common/src/cost_model.rs

```diff
@@ -0,0 +1,465 @@
+// Copyright 2023-, Edge & Node, GraphOps, and Semiotic Labs.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use sqlx::PgPool;
+use std::{collections::HashSet, str::FromStr};
+use thegraph_core::{DeploymentId, ParseDeploymentIdError};
+
+/// Internal cost model representation as stored in the database.
+///
+/// These can have "global" as the deployment ID.
+#[derive(Debug, Clone)]
+pub(crate) struct DbCostModel {
+    pub deployment: String,
+    pub model: Option<String>,
+    pub variables: Option<Value>,
+}
+
+/// External representation of cost models.
+///
+/// Here, any notion of "global" is gone and deployment IDs are valid deployment IDs.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CostModel {
+    pub deployment: DeploymentId,
+    pub model: Option<String>,
+    pub variables: Option<Value>,
+}
+
+impl TryFrom<DbCostModel> for CostModel {
+    type Error = ParseDeploymentIdError;
+
+    fn try_from(db_model: DbCostModel) -> Result<Self, Self::Error> {
+        Ok(Self {
+            deployment: DeploymentId::from_str(&db_model.deployment)?,
+            model: db_model.model,
+            variables: db_model.variables,
+        })
+    }
+}
+
+impl From<CostModel> for DbCostModel {
+    fn from(model: CostModel) -> Self {
+        let deployment = model.deployment;
+        DbCostModel {
+            deployment: format!("{deployment:#x}"),
+            model: model.model,
+            variables: model.variables,
+        }
+    }
+}
+
+/// Query cost models from the database, merging the global cost model in
+/// whenever there is no cost model defined for a deployment.
+pub async fn cost_models(
```

Comment:
In my opinion, this function does way too many different things. I recommend creating helper functions such as:

* `convert_deployments_to_hex`
* `query_cost_models`
* `filter_deployments`
* `merge_cost_model`

Reply:
Hey, I was just moving code from `indexer-service` to here to access this code.